module/state_synchronization/requester/oneshot_execution_data_requester.go

```diff
@@ -0,0 +1,182 @@
+				Uint64("height", blockHeight).
+				Uint64("attempt", uint64(attempt)).
+				Msgf("retrying download")
+
+			r.metrics.FetchRetried()
+		}
+		attempt++
+
+		// download execution data for the block
+		fetchTimeout, _ := timeout.Next()
+		var err error
+		execData, err = r.processFetchRequest(ctx, blockHeight, fetchTimeout)
+		if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) {
+			return retry.RetryableError(err)
+		}
+
+		if execution_data.IsMalformedDataError(err) || execution_data.IsBlobSizeLimitExceededError(err) {
+			// these errors indicate the execution data was received successfully and its hash matched
+			// the value in the ExecutionResult, however, the data was malformed or invalid. this means that
+			// an execution node produced an invalid execution data blob, and verification nodes approved it
+			r.log.Error().Err(err).Str(
+				"execution_data_id",
+				executionDataID.String(),
+			).Msg("received invalid execution data from network (potential slashing evidence?)")
+		}
+
+		return err
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return execData, nil
+}
+
+// processFetchRequest performs the actual fetch of execution data for the given block within the provided timeout.
+//
+// It wraps the fetch with metrics tracking and contextual logging. The execution data is retrieved
+// from the execution data cache based on the block ID.
+//
+// Expected errors during normal operations:
+// - storage.ErrNotFound if a seal or execution result is not available for the block
+// - BlobNotFoundError if some CID in the blob tree could not be found from the blobstore
+// - MalformedDataError if some level of the blob tree cannot be properly deserialized
+// - BlobSizeLimitExceededError if some blob in the blob tree exceeds the maximum allowed size
+// - context.DeadlineExceeded if fetching time exceeded fetchTimeout duration
+func (r *OneshotExecutionDataRequester) processFetchRequest(
+	parentCtx context.Context,
+	height uint64,
+	fetchTimeout time.Duration,
+) (*execution_data.BlockExecutionDataEntity, error) {
+	blockID := r.executionResult.BlockID
+	executionDataID := r.executionResult.ExecutionDataID
+
+	logger := r.log.With().
+		Str("block_id", blockID.String()).
+		Str("execution_data_id", executionDataID.String()).
+		Uint64("height", height).
+		Logger()
+	logger.Debug().Msg("processing fetch request")
+
+	start := time.Now()
+	r.metrics.ExecutionDataFetchStarted()
+	logger.Debug().Msg("downloading execution data")
+
+	ctx, cancel := context.WithTimeout(parentCtx, fetchTimeout)
+	defer cancel()
+
+	// NOTE: ByID does not add execData to cache or check if it is already in a cache, it only returns execData
+	execData, err := r.execDataCache.ByID(ctx, executionDataID)
+	r.metrics.ExecutionDataFetchFinished(time.Since(start), err == nil, height)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info().
+		Hex("execution_data_id", logging.ID(execData.ID())).
+		Msg("execution data fetched")
```

Comment:
`execution_data_id` is already included in `logger`
```suggestion
	logger.Info().Msg("execution data fetched")
```